### ISSUE TITLE
Making server.options.namingFunction async

### ISSUE THREAD

**Issue by Pitu** [tus/tus-node-server#421]
I want to name the incoming file based on a database check for collisions but it doesn't seem possible since the `namingFunction` is sync and does not allow to return the value of a resolved promise.

```ts
server.options.namingFunction = async req => {
	return await getUniqueFileIdentifier();
};
```

Another approach I tried was to call `getUniqueFileIdentifier()` inside `server.options.onUploadFinish`, and then rename the file accordingly, but the problem with this is that the tus-js-client makes a few more calls to the API after renaming the file and it can't find it, making it upload it again. 

So to summarize my objective is naming the file before the connection is closed so that I can return the filename to the user, but the function to name it needs to make a database call and I'm not to sure how to do that in this context.

**Comment by Pitu** [comment]
The way I solved it is by passing data around the metadata field the following way:

```ts
server.options.namingFunction = req => {
	return randomFilename();
};

server.options.onUploadCreate = async (req, res, upload) => {
	// Do some stuff with the upload
}

server.options.onUploadFinish = async (req, res, upload) => {
	const uniqueIdentifier = await getUniqueFileIdentifier();
	const newFileName = uniqueIdentifier + path.extname(upload.metadata.filename);
	upload.metadata.identifier = newFileName;
	// ...
	// return the identifier to the user
}

server.on(EVENTS.POST_FINISH, async (req, res, upload) => {
	// now that the connection is closed we can move the file without errors
	await jetpack.renameAsync(
		path.join(__dirname, 'uploads', upload.id),
		path.join(__dirname, 'uploads', upload.metadata.identifier)
	);
}
```

I don't know if this is the best approach, but it's working.
If you have any suggestions they are more than welcome.

**Comment by Murderlon** [comment]
Hi, just to check, could this be an [XY problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378) in which you try to check for duplicate file names, but another solution could be to write a `namingFunction` that is always unique (like the default)?
